internal/host/plugin/repository_host_set.go

```diff
@@ -152,6 +156,364 @@ func (r *Repository) CreateSet(ctx context.Context, scopeId string, s *HostSet,
+	}
+	if scopeId == "" {
+		return nil, nil, nil, db.NoRowsAffected, errors.New(ctx, errors.InvalidParameter, op, "no scope id")
+	}
+	if len(fieldMask) == 0 {
+		return nil, nil, nil, db.NoRowsAffected, errors.New(ctx, errors.EmptyFieldMask, op, "empty field mask")
+	}
+
+	// Get the old set first. We patch the record first before
+	// sending it to the DB for updating so that we can run on
+	// OnUpdateSet. Note that the field masks are still used for
+	// updating.
+	sets, plg, err := r.getSets(ctx, s.PublicId, "")
+	if err != nil {
+		return nil, nil, nil, db.NoRowsAffected, errors.Wrap(ctx, err, op)
+	}
+
+	if len(sets) == 0 {
+		return nil, nil, nil, db.NoRowsAffected, errors.New(ctx, errors.RecordNotFound, op, fmt.Sprintf("host set id %q not found", s.PublicId))
+	}
+
+	if len(sets) != 1 {
+		return nil, nil, nil, db.NoRowsAffected, errors.New(ctx, errors.Internal, op, fmt.Sprintf("unexpected amount of sets found, want=1, got=%d", len(sets)))
+	}
+
+	currentSet := sets[0]
+
+	// Assert the version of the current set to make sure we're
+	// updating the correct one.
+	if currentSet.GetVersion() != version {
+		return nil, nil, nil, db.NoRowsAffected, errors.New(ctx, errors.VersionMismatch, op, fmt.Sprintf("set version mismatch, want=%d, got=%d", currentSet.GetVersion(), version))
+	}
+
+	// Clone the set so that we can set fields.
+	newSet := currentSet.clone()
+	var updateAttributes bool
+	var dbMask, nullFields []string
+	const (
+		endpointOpNoop   = "endpointOpNoop"
+		endpointOpDelete = "endpointOpDelete"
+		endpointOpUpdate = "endpointOpUpdate"
+	)
+	var endpointOp string = endpointOpNoop
+	for _, f := range fieldMask {
+		switch {
+		case strings.EqualFold("name", f) && s.Name == "":
+			nullFields = append(nullFields, "name")
+			newSet.Name = s.Name
+		case strings.EqualFold("name", f) && s.Name != "":
+			dbMask = append(dbMask, "name")
+			newSet.Name = s.Name
+		case strings.EqualFold("description", f) && s.Description == "":
+			nullFields = append(nullFields, "description")
+			newSet.Description = s.Description
+		case strings.EqualFold("description", f) && s.Description != "":
+			dbMask = append(dbMask, "description")
+			newSet.Description = s.Description
+		case strings.EqualFold("PreferredEndpoints", f) && len(s.PreferredEndpoints) == 0:
+			endpointOp = endpointOpDelete
+			newSet.PreferredEndpoints = s.PreferredEndpoints
+		case strings.EqualFold("PreferredEndpoints", f) && len(s.PreferredEndpoints) != 0:
+			endpointOp = endpointOpUpdate
+			newSet.PreferredEndpoints = s.PreferredEndpoints
+		case strings.EqualFold("attributes", strings.Split(f, ".")[0]):
+			// Flag attributes for updating. While multiple masks may be
+			// sent, we only need to do this once.
+			updateAttributes = true
+
+		default:
+			return nil, nil, nil, db.NoRowsAffected, errors.New(ctx, errors.InvalidFieldMask, op, fmt.Sprintf("invalid field mask: %s", f))
+		}
+	}
+
+	if updateAttributes && s.Attributes != nil {
```

Comment:
if s.Attributes is nil but updateAttributes is true, then we should remove all attributes from this host set since that means the body of the post was something like
```
{
   "attributes": null
}
```

Reply:
Done!